data-prepper-plugins/common/src/main/java/org/opensearch/dataprepper/plugins/metricpublisher/MicrometerMetricPublisher.java

```diff
@@ -0,0 +1,74 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class MicrometerMetricPublisher implements MetricPublisher {
+    private static final String SERVICE_ID = "ServiceId";
+    private static final String OPERATION_NAME = "OperationName";
+    private static final Set<SdkMetric<String>> DEFAULT_DIMENSIONS = Stream.of(CoreMetric.SERVICE_ID, CoreMetric.OPERATION_NAME)
+            .collect(Collectors.toSet());
+
+    private final PluginMetrics pluginMetrics;
+
+    public MicrometerMetricPublisher(final PluginMetrics pluginMetrics) {
+        this.pluginMetrics = pluginMetrics;
+    }
+
+    @Override
+    public void publish(final MetricCollection metricCollection) {
+        final Map<String, String> dimensions = metricCollection.stream()
+                .filter(metricRecord -> DEFAULT_DIMENSIONS.contains(metricRecord.metric()))
+                .collect(Collectors.toMap(metricRecord -> metricRecord.metric().name(), metricRecord -> metricRecord.value().toString()));
+
+        addMetricsToPluginMetrics(metricCollection, dimensions);
+        metricCollection.children().forEach(
+                child -> addMetricsToPluginMetrics(child, dimensions)
+        );
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private void addMetricsToPluginMetrics(final MetricCollection metricCollection, final Map<String, String> dimensions) {
+        final String serviceIdValue = dimensions.get(SERVICE_ID);
+        final String operationNameValue = dimensions.get(OPERATION_NAME);
+
+        metricCollection.forEach(metricRecord -> {
+            final Class<?> metricValueClass = metricRecord.metric().valueClass();
+            final String metricName = metricRecord.metric().name();
+
+            if (Duration.class.isAssignableFrom(metricValueClass)) {
+                final long metricValueInMillis = Duration.parse(metricRecord.value().toString()).toMillis();
+                pluginMetrics.timerWithTags(metricName, SERVICE_ID, serviceIdValue, OPERATION_NAME, operationNameValue)
+                        .record(metricValueInMillis, TimeUnit.MILLISECONDS);
+            } else if (Number.class.isAssignableFrom(metricValueClass)) {
+                final Number numberMetricValue = (Number) metricRecord.value();
+                pluginMetrics.counterWithTags(metricName, SERVICE_ID, serviceIdValue, OPERATION_NAME, operationNameValue)
+                        .increment(numberMetricValue.doubleValue());
+            } else if (Boolean.class.isAssignableFrom(metricValueClass)) {
+                final Boolean booleanValue = (Boolean) metricRecord.value();
+                final double metricValue = Boolean.TRUE.equals(booleanValue) ? 1.0 : 0.0;
+                pluginMetrics.counterWithTags(metricName, SERVICE_ID, serviceIdValue, OPERATION_NAME, operationNameValue)
+                        .increment(metricValue);
+            }
```

Comment:
It could be useful to have an else with a log statement

Reply:
There are metrics with String type which will be added as dimensions. So I decided not to have an else because we don't want to do anything with String class or log them. There will be a lot of logs.

Reply:
Good point, it seems odd to me that they include things like `AwsRequestId` as a "metric". Long term it may be best to have an allowlist of the metrics we care about and filter on that. I am good the current implementation

Reply:
It might not be bad to have an `else` with trace logging. That way a user could turn it on if they are having a specific issue with metrics.

Reply:
@asifsmohammed , One of the advantages of using logger's built-in variable substitution is that it doesn't need to happen when the logging isn't needed.

Also, `String.format` is one of the slower methods for building strings.

If you change this to the following, then it will not have to format the string at all when trace logging is not enabled. And it will be faster when it is enabled.

```
LOG.trace("Ignoring {} metric. Only metrics with value Class Duration, Number, and Boolean are published.", metricName);
```